src/assets/wise5/classroomMonitor/dataExport/strategies/AbstractComponentDataExportStrategy.ts

```diff
@@ -0,0 +1,309 @@
+import { Component } from '../../../common/Component';
+import { ComponentDataExportParams } from '../ComponentDataExportParams';
+import { ComponentRevisionCounter } from '../ComponentRevisionCounter';
+import { UserIdsAndStudentNames } from '../UserIdsAndStudentNames';
+import { AbstractDataExportStrategy } from './AbstractDataExportStrategy';
+
+export abstract class AbstractComponentDataExportStrategy extends AbstractDataExportStrategy {
+  abstract COMPONENT_TYPE: string;
+
+  canViewStudentNames: boolean;
+  columnNames: string[] = [];
+  columnNameToNumber: Map<string, number> = new Map<string, number>();
+  includeOnlySubmits: boolean;
+  includeStudentNames: boolean;
+  workSelectionType: string;
+
+  constructor(protected component: Component, additionalParams: ComponentDataExportParams) {
+    super();
+    this.canViewStudentNames = additionalParams.canViewStudentNames;
+    this.includeOnlySubmits = additionalParams.includeOnlySubmits;
+    this.includeStudentNames = additionalParams.includeStudentNames;
+    this.workSelectionType = additionalParams.workSelectionType;
+  }
+
+  abstract generateComponentWorkRow(
+    rowCounter: number,
+    workgroupId: number,
+    userId1: number,
+    userId2: number,
+    userId3: number,
+    studentName1: string,
+    studentName2: string,
+    studentName3: string,
+    periodName: string,
+    componentRevisionCounter: ComponentRevisionCounter,
+    componentState: any
+  ): string[];
+
+  generateComponentHeaderRow(columnNames: string[]): string[] {
+    return [...columnNames];
+  }
+
+  populateColumnNames(): void {
+    const defaultColumnNames = [
+      '#',
+      'Workgroup ID',
+      'User ID 1',
+      'Student Name 1',
+      'User ID 2',
+      'Student Name 2',
+      'User ID 3',
+      'Student Name 3',
+      'Class Period',
+      'Project ID',
+      'Project Name',
+      'Run ID',
+      'Start Date',
+      'End Date',
+      'Server Timestamp',
+      'Client Timestamp',
+      'Node ID',
+      'Component ID',
+      'Component Part Number',
+      'Step Title',
+      'Component Type',
+      'Component Prompt',
+      'Student Data',
+      'Component Revision Counter',
+      'Is Submit',
+      'Submit Count'
+    ];
+    for (let c = 0; c < defaultColumnNames.length; c++) {
+      const defaultColumnName = defaultColumnNames[c];
+      this.columnNameToNumber.set(defaultColumnName, c);
+      this.columnNames.push(defaultColumnName);
+    }
+  }
+
+  generateComponentWorkRows(component: Component): string[] {
+    let rows = [];
+    let rowCounter = 1;
+    for (const workgroupId of this.configService.getClassmateWorkgroupIds()) {
+      const rowsForWorkgroup = this.generateComponentWorkRowsForWorkgroup(
+        workgroupId,
+        component.nodeId,
+        component.id,
```

Comment:
Is there a need to separate the component to nodeId and componentId and pass in two params here, or should we let the callee handle that?
```suggestion
        component,
```